### PR TITLE
Add a solution for Homework 4

### PR DESCRIPTION
**Pull request by tzurowski**
All test passed.

### CODE REVIEW

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            {
+                return weight / (float)Math.Pow(height, 2);
+            }
+            else
+            {
+                if (height <= 0 && weight <= 0)
+                    DisplayErrorBmi($"Weight cannot be equal or less than zero, but was {weight}.{Environment.NewLine}Height cannot be less than zero, but was {height}.");
+                else if(height <= 0)
+                    DisplayErrorBmi($"Height cannot be equal or less than zero, but was {height}.");
+                else
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'else' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        public static float PromptFloat(string message)
+        {
+            Console.WriteLine(message);
+
+            string input = Console.ReadLine();
+
+            if (IsValidNumber(input))
+                return float.Parse(input);
+            else if (String.IsNullOrEmpty(input))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            if (isValidHeightAndWeight)
+            {
+                return weight / (float)Math.Pow(height, 2);
+            }
+            else
+            {
+                if (height <= 0 && weight <= 0)
+                    DisplayErrorBmi($"Weight cannot be equal or less than zero, but was {weight}.{Environment.NewLine}Height cannot be less than zero, but was {height}.");
+                else if(height <= 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            bool isValidHeightAndWeight = height > 0 && weight > 0;
+
+            if (isValidHeightAndWeight)
+            {
+                return weight / (float)Math.Pow(height, 2);
+            }
+            else
+            {
+                if (height <= 0 && weight <= 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove this unnecessary 'using'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            Console.WriteLine($"Failed calculating BMI. Reason:{Environment.NewLine}{message}");
+        }
+
+        public static float PromptFloat(string message)
+        {
+            Console.WriteLine(message);
+
+            string input = Console.ReadLine();
+
+            if (IsValidNumber(input))
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            return IsNumber(input) && IsPositive(input);
+        }
+        public static bool IsNumber(string input)
+        {
+            return float.TryParse(input, out float number);
+        }
+        public static bool IsPositive(string input)
+        {
+            return (float.Parse(input) > 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs (hunk 1 above)
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: Added curly braces

File: Src/BootCamp.Chapter/Lesson4.cs (hunk 3 above)
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: Added curly braces

File: Src/BootCamp.Chapter/Lesson4.cs (hunk 5 above)
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: Added curly braces

File: Src/BootCamp.Chapter/Lesson4.cs (hunk 7 above)
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: Added curly braces

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            Console.WriteLine(message);
+
+            string name = Console.ReadLine();
+
+            if (String.IsNullOrWhiteSpace(name))
+            {
+                Console.Write("Name cannot be empty.");
+                return "-";
+            }
+            else
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'else' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: Added curly braces

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            return IsNumber(input) && IsPositive(input);
+        }
+        public static bool IsNumber(string input)
+        {
+            return float.TryParse(input, out float number);
+        }
+        public static bool IsPositive(string input)
+        {
+            return float.Parse(input) > 0;
+        }
```
Comment (ellmaster) [inline Src/BootCamp.Chapter/Lesson4.cs]: That is a lot of parsing just to find out that this is a valid number. Very inperformant. You can do this much nicer! :)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            string name = PromptNameString("What is your name?");
+            float age = PromptFloat("How old are you?");
+            float weight = PromptFloat("What is your weight (in kg)?");
+            float height = PromptFloat("What is your height (in cm)?");
+
+            Console.WriteLine($"{name} is {age} years old, his weight is {weight} kg and his height is {height} cm.");
+
+            float heightInMeters = height / 100;
+            float Bmi = CalculateBmi(weight, heightInMeters);
```
[the remark this answers is not in the document]
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: you are right, changed!

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+            if (isValidHeightAndWeight)
+            {
+                return weight / (float)Math.Pow(height, 2);
+            }
+            else
+            {
+                if (height <= 0 && weight <= 0)
+                {
+                    DisplayErrorBmi($"Weight cannot be equal or less than zero, but was {weight}.{Environment.NewLine}Height cannot be less than zero, but was {height}.");
```
[the remark this answers is not in the document]
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: I think it looks better now

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                return name;
+            }
+        }
+
+        public static float CalculateBmi(float weight, float height)
+        {
+            bool isValidHeightAndWeight = height > 0 && weight > 0;
+            string heightError = "Height cannot be equal or less than zero, but was ";
+            string heightErrorForNegative = "Height cannot be less than zero, but was ";
+            string weightError = "Weight cannot be equal or less than zero, but was ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'weightError'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            else
+            {
+                return name;
+            }
+        }
+
+        public static float CalculateBmi(float weight, float height)
+        {
+            bool isValidHeightAndWeight = height > 0 && weight > 0;
+            string heightError = "Height cannot be equal or less than zero, but was ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'heightError'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+            string heightErrorForNegative = "Height cannot be less than zero, but was ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'heightErrorForNegative'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs (hunk 10 above)
Comment (tzurowski) [reply Src/BootCamp.Chapter/Lesson4.cs]: I tried do it, but probably I done it worse than previous

File: Src/BootCamp.Chapter/Checks.cs [hunk 16]
```diff
@@ -16,26 +16,22 @@ public static class Checks
     {
         public static int PromptInt(string message)
         {
-            // To do: call your implementation. 
-            return 0;
+            return (int)Lesson4.PromptFloat(message);
```
[the remark this answers is not in the document]
Comment (tzurowski) [reply Src/BootCamp.Chapter/Checks.cs]: Changed, added new method for that

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'heightErrorForNegative'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+        public static bool IsValidFloat(string input)
+        {
+            bool isNumber = float.TryParse(input, out float number);
+            bool isPositive = (number > 0) ? true : false;
+            return isNumber && isPositive;
+        }
+        public static bool IsValidInt(string input)
+        {
+            bool isNumber = int.TryParse(input, out int number);
+            bool isPositive = (number > 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+            else
+            {
+                return name;
+            }
+        }
+
+        public static float CalculateBmi(float weight, float height)
+        {
+            bool isValidHeightAndWeight = height > 0 && weight > 0;
+            string heightError = "Height cannot be equal or less than zero, but was ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'heightError'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+            string heightErrorForNegative = "Height cannot be less than zero, but was ";
+            string weightError = "Weight cannot be equal or less than zero, but was ";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Add the 'const' modifier to 'weightError'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+        public static bool IsValidFloat(string input)
+        {
+            bool isNumber = float.TryParse(input, out float number);
+            bool isPositive = (number > 0) ? true : false;
+            return isNumber && isPositive;
+        }
+        public static bool IsValidInt(string input)
+        {
+            bool isNumber = int.TryParse(input, out int number);
+            bool isPositive = (number > 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+        public static bool IsValidFloat(string input)
+        {
+            bool isNumber = float.TryParse(input, out float number);
+            bool isPositive = number > 0;
+            return isNumber && isPositive;
+        }
+        public static bool IsValidInt(string input)
+        {
+            bool isNumber = int.TryParse(input, out int number);
+            bool isPositive = (number > 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

File: Src/BootCamp.Chapter/Lesson4.cs [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+        public static bool IsValidFloat(string input)
+        {
+            bool isValidNumber = float.TryParse(input, out float number) && (number > 0);
+            return isValidNumber;
+        }
+        public static bool IsValidInt(string input)
+        {
+            bool isNumber = int.TryParse(input, out int number);
+            bool isPositive = (number > 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Lesson4.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/Lesson4.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4971302&bid=16421964)

**Pull request closed without merging**